### PR TITLE
SONARPY-2014 SONARPY-2377 Repair rules reading PythonVisitorContext.pythonFile.contents() on Jupyter notebooks

### PR DESCRIPTION
**Pull request by ghislainpiot**
[SONARPY-2377](https://sonarsource.atlassian.net/browse/SONARPY-2377)

SONARPY-2014 TrailingWhitespaceCheck

[SONARPY-2377]: https://sonarsource.atlassian.net/browse/SONARPY-2377?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by joke1196** [comment]
Small reminder to rebase master as there are changes in the notebook parser test

**Comment by joke1196** [comment]
We should maybe create a ticket to refactor `.pythonLine` to have a specific type so that the user will not use `.line` instead of `.pythonLine`

**Review by joke1196 — CHANGES_REQUESTED** [review]
I think this looks good, as we mitigate a lot of issues. I think the next step is creating a few tickets. 
And for this PR we could maybe move the parser to the frontend?

File: python-frontend/src/main/java/org/sonar/plugins/python/api/IssueLocation.java [hunk 1]
```diff
@@ -38,7 +42,23 @@ public static IssueLocation atFileLevel(String message) {
...
+  public static IssueLocation atLineLevel(String message, int lineNumber, PythonFile pythonInputFile) {
+    if (pythonInputFile instanceof SonarQubePythonFile.IpynbFile ipynbFile) {
+      var generatedPythonFile = ipynbFile.pythonInputFile();
+      var mapping = generatedPythonFile.locationMap();
+      var begin = mapping.get(lineNumber);
+      Path path = pathOf(ipynbFile);
+      String fileId = path != null ? path.toString() : ipynbFile.toString();
+      LocationInFile locationInFile;
+      if (begin.isCompresssed()) {
+        var next = mapping.get(lineNumber + 1);
```
[the remark this answers is not in the document]
Comment (ghislainpiot) [reply python-frontend/src/main/java/org/sonar/plugins/python/api/IssueLocation.java:54]: What would we do if `next` is null? I am not sure it is recoverable

**Review by joke1196 — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**